packages/server/src/data/votes-data.js

```diff
@@ -0,0 +1,163 @@
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @returns {Promise<Models.Vote>}
+   * @public
+   */
+  async upvote(agentId, measurementId) {
+    return this._changeVote(agentId, measurementId, this._up);
+  }
+
+  /**
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @returns {Promise<Models.Vote>}
+   * @public
+   */
+  async downvote(agentId, measurementId) {
+    return this._changeVote(agentId, measurementId, this._down);
+  }
+
+  /**
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @param {Function} change
+   * @returns {Promise<Models.Vote>}
+   * @public
+   */
+  async _changeVote(agentId, measurementId, change) {
+    assert(!!agentId, 'Agent ID is required.');
+    assert(!!measurementId, 'Measurement ID is required.');
+
+    let transaction;
+    try {
+      transaction = await this.getTransaction();
+    } catch (e) {
+      this.log.trace(e.message, e);
+      throw e;
+    }
+
+    try {
+      const vote = await this._voteLocked(agentId, measurementId, transaction);
+      const voteAmount = change(vote);
+      if (voteAmount !== vote.voteAmount) {
+        await this._update(vote, voteAmount, transaction);
+      }
+      await this.commit(transaction);
+      return this._vote(agentId, measurementId, transaction);
+    } catch (e) {
+      this.log.trace(e.message, e);
+      await this.rollback(transaction);
+      throw e;
+    }
+  }
+
+  /**
+   * @param vote
+   * @returns {number}
+   * @protected
+   */
+  _up(vote) {
+    let voteAmount = vote.voteAmount;
+    if (voteAmount < 0) voteAmount = 0;
+    if (voteAmount <= 9) voteAmount += 1;
+    return voteAmount;
+  }
+
+  /**
+   * @param vote
+   * @returns {number}
+   * @protected
+   */
+  _down(vote) {
+    let voteAmount = vote.voteAmount;
+    if (voteAmount > 0) voteAmount = 0;
+    if (voteAmount >= -9) voteAmount -= 1;
+    return voteAmount;
+  }
+
+  /**
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @param {Layers.Transaction} transaction
+   * @returns {Promise<*>}
+   * @protected
+   */
+  async _voteLocked(agentId, measurementId, transaction) {
+    const params = new Params({ agentId, measurementId });
+    const query = new Query();
+    const lock = { level: transaction.LOCK.UPDATE };
+    const data = new Data({ agentId, measurementId });
+    const options = new Options({ transaction, lock, raw: true });
+    return this.upsertOne(params, query, data, options);
+  }
+
+  /**
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @param {Layers.Transaction} transaction
+   * @returns {Promise<*>}
+   * @protected
+   */
+  async _vote(agentId, measurementId, transaction) {
+    const params = new Params({ agentId, measurementId });
+    const query = new Query();
+    const options = new Options({ transaction, raw: true });
```

Comment:
If the vote is 0, do you think we should delete the vote? A vote of 0 should act identical to the lack of any vote.

Reply:
Since we just sum all votes it is not necessary to remove the vote.

Reply:
Sure, I guess the cost should be low.